File: client-typescript/src/hooks.ts
import type { Message, SessionEvent, ToolUse, ToolResult } from "./types.js";

/**
 * Optional harness customization points. Each hook receives the relevant,
 * mutable event and may inspect, mutate, or log it. Hooks may be async. A hook
 * that **throws** aborts the loop (surfaced as a {@link HookError}).
 *
 * Fire order per iteration of the loop:
 *   before_send → (send) → on_event* → after_receive →
 *   [ per tool_use:  before_tool_call → handler → after_tool_call ]
 */
export interface Hooks {
  /** Outgoing turn, just before it is sent. Mutations are sent. */
  before_send?: (message: Message) => void | Promise<void>;
  /** Assistant turn, immediately after it is received. */
  after_receive?: (message: Message) => void | Promise<void>;
  /** A `tool_use` request, before its handler is dispatched. */
  before_tool_call?: (toolUse: ToolUse) => void | Promise<void>;
  /**
   * A tool result, before it is sent back to the server. Rewriting
   * `toolResult.content` changes what the server receives.
   */
  after_tool_call?: (toolResult: ToolResult) => void | Promise<void>;
  /**
   * A live `session.event` observed on the `/rpc` notification stream during a
   * turn, in arrival order. Read-only: the same events are also available in
   * bulk from the terminal result, so agents that ignore live progress can
   * leave this unset.
   */
  on_event?: (event: SessionEvent) => void | Promise<void>;
}

/** The four hook point names, for diagnostics. */
export type HookName = keyof Hooks;

File: client-python/src/bae_py/hooks.py
"""Optional lifecycle hooks. Each hook receives the relevant event object, may
mutate it in place or log it, and may raise to abort the loop (surfaced as a
:class:`~bae_py.errors.HookError`).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Optional, Union

from .types import Message, SessionEvent, ToolResultBlock, ToolUseBlock

# A hook is any callable taking the event; return value is ignored. Sync or
# async — the harness awaits it if it is awaitable.
Hook = Callable[[Any], Union[None, Awaitable[None]]]


@dataclass(slots=True)
class Hooks:
    """The harness hook points (all optional).

    * ``before_send`` — an outgoing :class:`Message` about to be sent
      (fires for every turn, including the tool-result turn).
    * ``after_receive`` — the assistant :class:`Message` just received.
    * ``before_tool_call`` — a :class:`ToolUseBlock` about to be dispatched.
    * ``after_tool_call`` — the :class:`ToolResultBlock` produced by a handler,
      before it is sent back (rewriting ``content`` changes what is sent).
    * ``on_event`` — a live :class:`SessionEvent` observed on the ``/rpc``
      notification stream during a turn, in arrival order (read-only; the same
      events are also available in bulk from :attr:`Session.last_events`).
    """

    before_send: Optional[Callable[[Message], Union[None, Awaitable[None]]]] = None
    after_receive: Optional[Callable[[Message], Union[None, Awaitable[None]]]] = None
    before_tool_call: Optional[Callable[[ToolUseBlock], Union[None, Awaitable[None]]]] = None
    after_tool_call: Optional[Callable[[ToolResultBlock], Union[None, Awaitable[None]]]] = None
    on_event: Optional[Callable[[SessionEvent], Union[None, Awaitable[None]]]] = None
